lib/cln.js

```diff
@@ -2,30 +2,44 @@ import fetch from 'cross-fetch'
 import crypto from 'crypto'
 import { getAgent } from '@/lib/proxy'
 import { assertContentTypeJson, assertResponseOk } from './url'
+import { FetchTimeoutError } from './fetch'
+import { WALLET_CREATE_INVOICE_TIMEOUT_MS } from './constants'
 
-export const createInvoice = async ({ msats, description, expiry }, { socket, rune, cert }) => {
+export const createInvoice = async ({ msats, description, expiry }, { socket, rune, cert }, { signal }) => {
   const agent = getAgent({ hostname: socket, cert })
 
   const url = `${agent.protocol}//${socket}/v1/invoice`
-  const res = await fetch(url, {
-    method: 'POST',
-    headers: {
-      'Content-Type': 'application/json',
-      Rune: rune,
-      // can be any node id, only required for CLN v23.08 and below
-      // see https://docs.corelightning.org/docs/rest#server
-      nodeId: '02cb2e2d5a6c5b17fa67b1a883e2973c82e328fb9bd08b2b156a9e23820c87a490'
-    },
-    agent,
-    body: JSON.stringify({
-      // CLN requires a unique label for every invoice
-      // see https://docs.corelightning.org/reference/lightning-invoice
-      label: crypto.randomBytes(16).toString('hex'),
-      description,
-      amount_msat: msats,
-      expiry
+
+  let res
+  try {
+    res = await fetch(url, {
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/json',
+        Rune: rune,
+        // can be any node id, only required for CLN v23.08 and below
+        // see https://docs.corelightning.org/docs/rest#server
+        nodeId: '02cb2e2d5a6c5b17fa67b1a883e2973c82e328fb9bd08b2b156a9e23820c87a490'
+      },
+      agent,
+      body: JSON.stringify({
+        // CLN requires a unique label for every invoice
+        // see https://docs.corelightning.org/reference/lightning-invoice
+        label: crypto.randomBytes(16).toString('hex'),
+        description,
+        amount_msat: msats,
+        expiry
+      }),
+      signal
     })
-  })
+  } catch (err) {
+    if (err.name === 'AbortError') {
+      // XXX node-fetch doesn't throw our custom TimeoutError but throws a generic error so we have to handle that manually.
+      // see https://github.com/node-fetch/node-fetch/issues/1462
+      throw new FetchTimeoutError('POST', url, WALLET_CREATE_INVOICE_TIMEOUT_MS)
```

Comment:
this assumes the signal was a timeout signal that used `WALLET_CREATE_INVOICE_TIMEOUT_MS` as the timeout which isn't clear from the immediate context but since we are smart and know more than just the immediate context and this is a workaround for https://github.com/node-fetch/node-fetch/issues/1462, I think this is okay